ChildProject/pipelines/metrics.py

```diff
@@ -0,0 +1,391 @@
+from abc import ABC, abstractmethod
+import argparse
+import datetime
+import multiprocessing as mp
+import numpy as np
+import pandas as pd
+from typing import Union, List
+
+import ChildProject
+from ChildProject.pipelines.pipeline import Pipeline
+
+pipelines = {}
+
+class Metrics(ABC):
+    def __init__(self,
+        project: ChildProject.projects.ChildProject,
+        by: str = 'recording_filename',
+        recordings: Union[str, List[str], pd.DataFrame] = None):
+
+        self.project = project
+        self.am = ChildProject.annotations.AnnotationManager(self.project)
+
+        recording_columns = {
+            'recording_filename', 'child_id', 'duration', 'session_id', 'session_offset'
+        }
+        recording_columns &= set(self.project.recordings.columns)
+
+        self.am.annotations = self.am.annotations.merge(
+            self.project.recordings[recording_columns],
+            left_on = 'recording_filename',
+            right_on = 'recording_filename'
+        )
+
+        self.by = by
+        self.segments = pd.DataFrame()
+
+        if recordings is None:
+            self.recordings = None
+        elif isinstance(recordings, pd.DataFrame):
+            if 'recording_filename' not in recordings.columns:
+                raise ValueError("recordings dataframe is missing a 'recording_filename' column")
+            self.recordings = recordings['recording_filename'].tolist()
+        elif isinstance(recordings, pd.Series):
+            self.recordings = recordings.tolist()
+        elif isinstance(recordings, list):
+            self.recordings = recordings
+        else:
+            if not os.path.exists(recordings):
+                raise ValueError(
+                    "'recordings' is neither a pandas dataframe,"
+                    "nor a list or a path to an existing dataframe."
```

Comment:
can we give the user an instruction RE what to do about this?

Reply:
eg check parameters passed to the function, or check structure of dataframe against instructions LINK